src/routes/notifications/v1/notifications.controller.ts

```diff
@@ -11,40 +12,149 @@ import { RegisterDeviceDto } from '@/routes/notifications/v1/entities/register-d
 import { NotificationsService } from '@/routes/notifications/v1/notifications.service';
 import { ValidationPipe } from '@/validation/pipes/validation.pipe';
 import { AddressSchema } from '@/validation/entities/schemas/address.schema';
+import type {
+  UpsertSubscriptionsDto,
+  UpsertSubscriptionsSafesDto,
+} from '@/routes/notifications/v2/entities/upsert-subscriptions.dto.entity';
+import { AuthPayload } from '@/domain/auth/entities/auth-payload.entity';
+import { NotificationType } from '@/domain/notifications/v2/entities/notification.entity';
+import type { UUID } from 'crypto';
+import { NotificationsServiceV2 } from '@/routes/notifications/v2/notifications.service';
+import { recoverMessageAddress } from 'viem';
+import { UuidSchema } from '@/validation/entities/schemas/uuid.schema';
 
 @ApiTags('notifications')
 @Controller({ path: '', version: '1' })
 export class NotificationsController {
-  constructor(private readonly notificationsService: NotificationsService) {}
+  constructor(
+    // Adding NotificationServiceV2 to ensure compatibility with V1.
+    // @TODO Remove NotificationModuleV2 after all clients have migrated and compatibility is no longer needed.
+    @Inject(NotificationsServiceV2)
+    private readonly notificationServiceV2: NotificationsServiceV2,
+    private readonly notificationsService: NotificationsService,
+  ) {}
 
   @ApiOkResponse()
   @Post('register/notifications')
   @HttpCode(200)
   async registerDevice(
     @Body() registerDeviceDto: RegisterDeviceDto,
   ): Promise<void> {
-    return this.notificationsService.registerDevice(registerDeviceDto);
+    await this.notificationsService.registerDevice(registerDeviceDto);
+
+    // Compatibility with V2
+    // @TODO Remove NotificationModuleV2 after all clients have migrated and compatibility is no longer needed.
+    const compatibleV2Requests =
+      await this.createV2RegisterDto(registerDeviceDto);
+
+    const v2Requests = [];
+    for (const compatibleV2Request of compatibleV2Requests) {
+      v2Requests.push(
+        await this.notificationServiceV2.upsertSubscriptions(
+          compatibleV2Request,
+        ),
+      );
+    }
+
+    await Promise.all(v2Requests);
+  }
+
+  private async createV2RegisterDto(args: RegisterDeviceDto): Promise<
+    Array<{
+      upsertSubscriptionsDto: UpsertSubscriptionsDto;
+      authPayload: AuthPayload;
+    }>
+  > {
+    const safeV2Array: Array<{
+      authPayload: AuthPayload;
+      upsertSubscriptionsDto: UpsertSubscriptionsDto & {
+        signature: `0x${string}`;
+      };
+    }> = [];
+
+    const safesV1Registrations = args.safeRegistrations;
+
+    for (const safeV1Registration of safesV1Registrations) {
+      if (safeV1Registration.safes.length) {
+        const safeV2: {
+          authPayload: AuthPayload;
+          upsertSubscriptionsDto: UpsertSubscriptionsDto & {
+            safes: Array<UpsertSubscriptionsSafesDto>;
+            signature: `0x${string}`;
+          };
```

Comment:
I totally agree. Since we usually define types inline, I did it this way to keep it consistent. Do you think we should consider creating an external type for it?

Reply:
As we discussed, I find heavy use of utility types makes the code extremely hard to read. However, I’d be happy to make the change if you could provide a code snippet as an example.

Reply:
No strong opinion here, but I'd prefer to use utility types so we can have a SSOT as Aaron suggests 🙂 Feel free to keep the current implementation if you prefer.

Reply:
Done
https://github.com/safe-global/safe-client-gateway/pull/2063/commits/5e5999d1ec9eb3cd743b074e7869be41f8bf1808

Reply:
Since the PR is approved, I'm gonna resolve this thread.